download-basket/src/main/java/downloadbasket/helpers/OGCServices.java

```diff
@@ -0,0 +1,189 @@
+package downloadbasket.helpers;
+
+import fi.nls.oskari.log.LogFactory;
+import fi.nls.oskari.util.PropertyUtil;
+import downloadbasket.data.NormalWayDownloads;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.io.StringWriter;
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+
+public class OGCServices {
+	private static final fi.nls.oskari.log.Logger LOGGER = LogFactory.getLogger(OGCServices.class);
+
+	private static final String PARAM_CROPPING_MODE = "croppingMode";
+	private static final String PARAM_CROPPING_LAYER = "croppingLayer";
+	private static final String PARAM_BBOX = "bbox";
+	private static final String PARAM_BBOX_LEFT = "left";
+	private static final String PARAM_BBOX_BOTTOM = "bottom";
+	private static final String PARAM_BBOX_RIGHT = "right";
+	private static final String PARAM_BBOX_TOP = "top";
+	private static final String PARAM_IDENTIFIERS = "identifiers";
+	private static final String PARAM_LAYER = "layer";
+
+	/**
+	 * Get filter
+	 * 
+	 * @param downloadDetails
+	 *            download details
+	 * @param writeParam
+	 *            write param
+	 *
+	 * @return filter url param and value
+	 */
+	public static String getFilter(JSONObject downloadDetails, Boolean writeParam)
+			throws JSONException, UnsupportedEncodingException {
+		StringWriter s = new StringWriter();
+
+		String[] bboxDownloadTypes = PropertyUtil.getCommaSeparatedList("oskari.wfs.bbox.downloads");
+		NormalWayDownloads normalDownloads = new NormalWayDownloads();
+		for (String download : bboxDownloadTypes) {
+			normalDownloads.addDownload(download);
+		}
+
+		final String croppingMode = downloadDetails.getString(PARAM_CROPPING_MODE);
+		String croppingLayer = "";
+		if (downloadDetails.has(PARAM_CROPPING_LAYER)) {
+			croppingLayer = downloadDetails.getString(PARAM_CROPPING_LAYER);
+		}
+
+		if (normalDownloads.isBboxCropping(croppingMode, croppingLayer)) {
+			if (writeParam) {
+				s.append("&bbox=");
+			}
+			s.append(getBbox(downloadDetails.getJSONObject(PARAM_BBOX)));
+		} else {
+			if (writeParam) {
+				s.append("&filter=");
+			}
+			s.append(URLEncoder.encode(getPluginFilter(downloadDetails, true, false), "UTF-8"));
+		}
+
+		return s.toString();
+	}
+
+	/**
+	 * Get bbox
+	 * 
+	 * @param bbox
+	 * @return
+	 * @throws JSONException
+	 */
+	private static String getBbox(JSONObject bbox) throws JSONException {
+		return bbox.getString(PARAM_BBOX_LEFT) + "," + bbox.getString(PARAM_BBOX_BOTTOM) + ","
+				+ bbox.getString(PARAM_BBOX_RIGHT) + "," + bbox.getString(PARAM_BBOX_TOP);
+	}
+
+	/**
+	 * Get WFS Query layer plugin WFS request
+	 * 
+	 * @param download
+	 *            download details
+	 * @param addNameSpaces
+	 *            add namespaces?
+	 * @param writeParam
+	 *            write param
+	 *
+	 * @return filter writer
+	 */
+	public static String getPluginFilter(JSONObject download, boolean addNameSpaces, boolean writeParam)
```

Comment:
This function is too complex. Manually creating the XML is also extremely error prone. Also the addNamespaces is not used consistently, for example L125-L129 always add the `ogc:` prefix to the elements. Please consider using a library for building the OGC Filters, for example [Geotools Filter](http://docs.geotools.org/latest/userguide/library/main/filter.html).